pkg/didcomm/protocol/decorator/decorator.go

```diff
@@ -118,7 +129,158 @@ func (d *AttachmentData) Fetch() ([]byte, error) {
 
 	// TODO add support to fetch links
 
-	// TODO add support for jws signatures
-
 	return nil, errors.New("no contents in this attachment")
 }
+
+type rawSig struct {
+	Header    rawSigHeader `json:"header,omitempty"`
+	Signature string       `json:"signature,omitempty"`
+	Protected string       `json:"protected,omitempty"`
+}
+
+type rawSigHeader struct {
+	KID string `json:"kid,omitempty"`
+}
+
+type rawProtected struct {
+	JWK json.RawMessage `json:"jwk,omitempty"`
+	Alg string          `json:"alg,omitempty"`
+}
+
+// Sign signs the base64 payload of the AttachmentData, and adds the signature to the attachment.
+func (d *AttachmentData) Sign(c crypto.Crypto, kh interface{}, pub interface{}, pubBytes []byte) error {
+	didKey, _ := fingerprint.CreateDIDKey(pubBytes)
+
+	jwk, err := jose.JWKFromKey(pub)
+	if err != nil {
+		return fmt.Errorf("creating jwk from pub key: %w", err)
+	}
+
+	jwk.Kty = "OKP"
+	jwk.KeyID = didKey
+
+	jwkBytes, err := jwk.MarshalJSON()
+	if err != nil {
+		return fmt.Errorf("marshaling jwk: %w", err)
+	}
+
+	protected := rawProtected{
+		JWK: jwkBytes,
+		Alg: "EdDSA",
+	}
+
+	protectedBytes, err := json.Marshal(protected)
+	if err != nil {
+		return fmt.Errorf("marshaling protected header: %w", err)
+	}
+
+	protectedB64 := base64.RawURLEncoding.EncodeToString(protectedBytes)
+
+	b64data := strings.Trim(d.Base64, "=")
+
+	signedData := fmt.Sprintf("%s.%s", protectedB64, b64data)
+
+	sig, err := c.Sign([]byte(signedData), kh)
+	if err != nil {
+		return fmt.Errorf("signing data: %w", err)
+	}
+
+	jws := rawSig{
+		Header: rawSigHeader{
+			KID: didKey,
+		},
+		Protected: protectedB64,
+		Signature: base64.RawURLEncoding.EncodeToString(sig),
+	}
+
+	serializedJWS, err := json.Marshal(jws)
+	if err != nil {
+		return fmt.Errorf("marshaling jws: %w", err)
+	}
+
+	d.JWS = serializedJWS
+
+	return nil
+}
+
+// Verify verify the signature on the attachment data.
+func (d *AttachmentData) Verify(c crypto.Crypto, keyManager kms.KeyManager) error {
+	if d.JWS == nil {
+		return fmt.Errorf("no signature to verify")
+	}
+
+	jws := rawSig{}
+
+	err := json.Unmarshal(d.JWS, &jws)
+	if err != nil {
+		return fmt.Errorf("parsing jws: %w", err)
+	}
+
+	sigKey, err := fingerprint.PubKeyFromDIDKey(jws.Header.KID)
+	if err != nil {
+		return fmt.Errorf("parsing did:key '%s': %w", jws.Header.KID, err)
+	}
+
+	protectedBytes, err := base64.RawURLEncoding.DecodeString(jws.Protected)
+	if err != nil {
+		return fmt.Errorf("decoding protected header: %w", err)
+	}
+
+	var protected rawProtected
+
+	err = json.Unmarshal(protectedBytes, &protected)
+	if err != nil {
+		return fmt.Errorf("parsing protected header: %w", err)
+	}
+
+	jwk := jose.JWK{}
+
+	err = jwk.UnmarshalJSON(protected.JWK)
+	if err != nil {
+		return fmt.Errorf("parsing jwk: %w", err)
+	}
+
+	var keyType kms.KeyType
+
+	switch jwk.Key.(type) {
```

Comment:
@Baha-sk is this a good way to select the kms key type from the jwk? Is there a way that you've already created to do that?

Reply:
this whole code is duplicated, why not use afgo's doc/jose package instead?

Reply:
I'm already using doc/jose, and it doesn't seem to support JSON JWS, only compact, so I'm using it just for the JWKs. There isn't any mechanism in doc/jose.JWK for determining the `kms.KeyType`, and I didn't find anything in the kms either.

Reply:
so figuring out the kms key type from a JWK doesn't exist directly.

Currently doc/jose/jwk.go has internal functions to check the key type like [this existing code](
https://github.com/hyperledger/aries-framework-go/blob/main/pkg/doc/jose/jwk.go#L216-L260)

you can probably add a member function to JWK that returns the kms keyType based on the embedded key using these functions